### PR TITLE
[LibP2P] - Implement Request/Response For The PushCDN Network

### LINKED ISSUE #3565: [LibP2P] - Implement Request/Response For The PushCDN Network
### What is this task and why do we need to work on it?

Request/Response is not yet available for the PushCDN network layer. We need to implement this functionality to be able to route such requests through PushCDN. In particular, we need to implement the following specific trait methods:
- `spawn_request_receiver_task`
- `request_data`

These methods enable the processing of messages where arbitrary data is being fetched from a peer.

### `spawn_request_receiver_task`
This will require the creation of a broadcast channel for receiving messages from `NetworkResponseState` and processing. This `mpsc` channel will exist within the pushcdn type and be made available for processing requests forwarded to the background task.

### `request_data`
We'll use this method to actually facilitate requesting the designated peer.

This work will also remove the dependency on `TYPES` within the `request_data` method and instead put it into the implementor's domain. This is to get around an issue like the following:
```rust
#[async_trait]
impl<TYPES: NodeType> ConnectedNetwork<TYPES::SignatureKey> for PushCdnNetwork<TYPES> {
    async fn request_data<TYPES2: NodeType>(
        &self,
        request: Vec<u8>,
        recipient: TYPES2::SignatureKey,
    ) -> Result<Vec<u8>, NetworkError> {
        self.client.send_direct_message(recipient, request).await;

        Ok(vec![])
    }

    async fn spawn_request_receiver_task(
        &self,
    ) -> Option<mpsc::Receiver<(Vec<u8>, ResponseChannel<Vec<u8>>)>> {
        None
    }
```

Right now, a network backend that uses `TYPES` natively will not be able to cleanly implement the trait method as we need to perform disambiguation on the types. The alternative is to make `TYPES` a member of `ConnectedNetwork`, however, this requires tedious changes to the sequencer and, as a result, is less ideal.

### Are there any other details to include?

_No response_

### What are the acceptance criteria to close this issue?

- New implementation methods are added
- This backend is added to the test suite
- New tests are added for the new behavior

### Branch work will be merged to (if not the default branch)

_No response_

### PR DESCRIPTION
**Pull request by jparr721**
Closes #3565 

### This PR: 
Adds broad support for request/response for Quorum Proposals across the entire network layer. It accomplishes this by introducing new HotShot events that are able to interpret requests for data. It leverages the existing `NetworkRequestTask` to read events and respond if data exists.

This PR also updates the `NetworkTask` to support the new events, and supports direct messaging between the requesting and responding node.

### CODE REVIEW

File: crates/types/src/message.rs
Comment (rob-maron) [inline crates/types/src/message.rs:169]: This message is needed to "finish" the request right? Is there any downside to sending-and-forgetting the request and getting a proposal, and then doing normal consensus operations with it?

File: crates/testing/tests/tests_1/quorum_proposal_recv_task.rs
Comment (rob-maron) [inline crates/testing/tests/tests_1/quorum_proposal_recv_task.rs:216]: Not as familiar with the integration tests - why do we expect a request for view 2 here?

File: crates/task-impls/src/network.rs
Comment (rob-maron) [inline crates/task-impls/src/network.rs:312]: This is a future change - but I believe a malicious node could put a `sender_key` that's not them and amplify traffic to them, as they would get the response proposal

Perhaps we could sign over this

File: crates/task-impls/src/helpers.rs
Comment (rob-maron) [inline crates/task-impls/src/helpers.rs:81]: Similar to the comment in `message.rs` - we could get rid of the need for an extra task if we send and forget. But not sure how nicely that plays with the dependency stuff

File: crates/hotshot/src/traits/networking/push_cdn_network.rs
Comment (rob-maron) [inline crates/hotshot/src/traits/networking/push_cdn_network.rs:445]: Are we getting rid of these entirely in the network function signatures? Do we need them?

**Review by rob-maron — COMMENTED** [review]
Looks good, just some high-level questions for my understanding

File: crates/testing/tests/tests_1/quorum_proposal_recv_task.rs
Comment (jparr721) [reply crates/testing/tests/tests_1/quorum_proposal_recv_task.rs:216]: That proposal is missing in the parent's state. Looking here
```rust
    let inputs = vec![serial![QuorumProposalRecv(
        proposals[2].clone(),
        leaders[2]
    )]];
```

We were sent a proposal for view 3, but the proposal for view 2 (i.e. the `justify_qc`) does *not* exist in memory due to the setup of the test, so we broadcast a request for the proposal from the leader to get the proposal so we know that we can safely branch off of this node.

File: crates/task-impls/src/network.rs
Comment (jparr721) [reply crates/task-impls/src/network.rs:312]: Yeah I know we had chatted about this before. Tobias and I had talked about it as well but the crux of the matter is that anyone at any time could spam a single node by adding a new internal event that always broadcasts and roasts the leader. I think we'll have to just penalize to avoid spamming, even if we sign it to ensure validity I can still get bogus data through the network layer to whomever I want, even if they end up dropping the message.

File: crates/task-impls/src/helpers.rs
Comment (jparr721) [reply crates/task-impls/src/helpers.rs:81]: The downside of send-and-forget is that it disallows us the ability to block on the data coming back. So, in this case, the proposal's arrival begets a block before we vote/propose if we're missing that prior state. Because of that, I opted to just block (as the prior version did as well) and timeout in some short interval if we don't hear back from the leader in time. I think in general though, I agree that this type of blocking should *not* happen. I'd like to do a more earnest look at the R/R tasks but I didn't want to make this a massive PR.

File: crates/testing/tests/tests_1/quorum_proposal_recv_task.rs
Comment (rob-maron) [reply crates/testing/tests/tests_1/quorum_proposal_recv_task.rs:216]: Awesome, that's what I suspected, thanks!

File: crates/hotshot/src/traits/networking/push_cdn_network.rs
Comment (jparr721) [reply crates/hotshot/src/traits/networking/push_cdn_network.rs:445]: Unfortunately yes, but I'd like to delete these methods entirely. If you look in the `request.rs` file, the vid request uses the `DelayedRequester` which calls `request_data`. So, it'll fail for the PushCDN, but our backstop from this is that combined network *only* uses the LibP2p layer as-is anyway, so I don't think this'll cause problems. Admittedly, this particular solution is not my favorite. I plan to do essentially this exact same changeset, but for the VID data request, in a follow-on PR. I think we could wait to merge this and I can branch from this and do the VID change, merge that one into this one, then ship it all as one macro-feature if we think that this approach could cause breakage. Admittedly I was not sure.

File: crates/task-impls/src/network.rs
Comment (rob-maron) [reply crates/task-impls/src/network.rs:312]: Right, but I guess in this case you could essentially have all _other_ nodes DoS this one node, which might be harder to deal with

But long-term thing if any, not significant for this PR

File: crates/task-impls/src/helpers.rs
Comment (rob-maron) [reply crates/task-impls/src/helpers.rs:81]: Gotcha, yeah. Maybe down the line we could utilize the dependency framework to get it to block - although admittedly I am not the most familiar. For this PR it is fine though

File: crates/hotshot/src/traits/networking/push_cdn_network.rs
Comment (rob-maron) [reply crates/hotshot/src/traits/networking/push_cdn_network.rs:445]: Sounds good!

**Review by rob-maron — APPROVED** [review]

File: crates/types/src/message.rs
Comment (jparr721) [reply crates/types/src/message.rs:169]: Yeah so basically it is the receiving node getting a `QuorumProposalResponseSend` event, and transforming it into a `QuorumProposalResponseRecv` event. This could be done any number of ways, I just opted to stratify things a little bit here because I prefer to have explicit events for each stage of the request chain. This GeneralConsensusMessage type honestly only exists to fulfill the need for there to be a consensus message that will not 
A. Be confused with a genuine proposal.
B. Be confused with a request for a proposal.

I'd happily name this anything else if there's any opinions there.

On your send-and-forget remark, I think generally I would like this, but it comes with two caveats given our async-focused design in HotShot

1. Right now, if I need to **guarantee** that data is available in my local in-memory state, I can *only* spin lock in a background thread (with some arbitrary delay).
2. I need a way to be able to halt a critical operation until my data has arrived.

As a result, the event dependency made the most sense. So, yes, *technically* we for sure could send and forget to get the proposal, the problem is we'd end up invariably waiting anyway, so by making the events explicitly say "I want a proposal right now for view x and I will wait until it is found" we can specialize what they're doing.

**Review by bfish713 — APPROVED** [review]

File: crates/task-impls/src/network.rs
Comment (bfish713) [reply crates/task-impls/src/network.rs:312]: Actually previously the DataRequest message type was signed by the sender so this type of amplification/DDoS attack would not be possible, this PR is actually removing the signing.  We should added it back soon than later

File: crates/task-impls/src/request.rs
Comment (bfish713) [inline crates/task-impls/src/request.rs:126]: It's not really wrong to have this here, but it logically belongs in the "Response" task which handles responding to requests where the request task is handling sending out requests and waiting for the response

File: crates/task-impls/src/network.rs
Comment (bfish713) [inline crates/task-impls/src/network.rs:311]: Just FYI the next task related to this is to make it so we can request proposals from the DA committee as well so this logic will have to get significantly more complex.  Basically the logic of do_vid in the request task will need to live here now, or we can just handle this event there like before

**Pull request merged**
